File: components/languages/python/runner.py
#!/usr/bin/env python3
"""KubeSandbox batch runner for the Python component (doc §5.3).

Executes user code in an isolated namespace so the runner's own imports/locals never
leak into the captured variable dump, then serializes the final namespace to
VAR_DUMP_PATH for the provisioner to read after the process exits.

stdout/stderr are deliberately left untouched (inherited from this process) — the
provisioner captures those at the container level, not here. stdin is also inherited:
the provisioner writes the batch run's up-front stdin then closes the pipe (EOF) before
or at execution start, so a user `input()` call either gets real data or an immediate
EOFError — it never blocks waiting on a live client (doc §5.1).
"""

from __future__ import annotations

import json
import sys
import traceback
import types

VAR_DUMP_PATH = "/tmp/.kubesandbox_vars.json"


def _is_json_safe(value: object) -> bool:
    try:
        json.dumps(value)
        return True
    except TypeError:
        return False


def _dump_namespace(namespace: dict) -> dict:
    dumped: dict[str, object] = {}
    for name, value in namespace.items():
        if name.startswith("__"):
            continue
        if callable(value) or isinstance(value, types.ModuleType):
            continue
        dumped[name] = value if _is_json_safe(value) else repr(value)
    return dumped


def main() -> int:
    if len(sys.argv) < 2:
        print("usage: python_runner.py <file>", file=sys.stderr)
        return 2

    source_path = sys.argv[1]
    with open(source_path, encoding="utf-8") as fh:
        source = fh.read()

    namespace: dict[str, object] = {"__name__": "__main__", "__file__": source_path}
    exit_code = 0
    try:
        code = compile(source, source_path, "exec")
        exec(code, namespace)
    except SystemExit as exc:
        exit_code = exc.code if isinstance(exc.code, int) else (1 if exc.code else 0)
    except BaseException:
        traceback.print_exc()
        exit_code = 1
    finally:
        try:
            with open(VAR_DUMP_PATH, "w", encoding="utf-8") as fh:
                json.dump(_dump_namespace(namespace), fh)
        except OSError as dump_exc:
            # Still best-effort (a missing dump just yields variables=null upstream),
            # but silently swallowing this hid a real bug once already — never again.
            print(f"kubesandbox_runner: failed to write variable dump: {dump_exc}", file=sys.stderr)

    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: components/languages/node/runner.js
#!/usr/bin/env node
/**
 * KubeSandbox batch runner for the Node.js component (doc §5.3).
 *
 * Mirrors the Python runner's contract (isolated namespace, dump final scope to
 * VAR_DUMP_PATH) with one real limitation worth being explicit about: `vm.createContext`
 * makes a plain object act as the script's global object, and top-level `var` (or an
 * implicit global assignment like `x = 1`) becomes a property of that object — but
 * top-level `let`/`const` bindings live in the script's lexical environment, which V8
 * does not expose as an inspectable object through any public Node API. So unlike
 * Python's exec()-namespace dump (which captures every top-level assignment regardless
 * of how it was declared), this runner only ever captures `var`/implicit-global
 * bindings. That's a real JS-semantics constraint, not a bug — documented here instead
 * of silently under-delivering on doc §5.3's "global variable dump" framing.
 *
 * stdout/stderr/stdin are left to this process (inherited) — the provisioner captures
 * stdout/stderr at the container level and feeds/EOFs stdin up front (doc §5.1), same
 * as the Python runner.
 */

'use strict';

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const VAR_DUMP_PATH = '/tmp/.kubesandbox_vars.json';

function isJsonSafe(value) {
  try {
    JSON.stringify(value);
    return true;
  } catch (err) {
    return false;
  }
}

function dumpContext(sandbox, ownKeys) {
  const dumped = {};
  for (const name of ownKeys) {
    if (name.startsWith('__')) continue;
    const value = sandbox[name];
    if (typeof value === 'function') continue;
    dumped[name] = isJsonSafe(value) ? value : String(value);
  }
  return dumped;
}

function main() {
  const sourcePath = process.argv[2];
  if (!sourcePath) {
    process.stderr.write('usage: node_runner.js <file>\n');
    return 2;
  }

  const source = fs.readFileSync(sourcePath, 'utf-8');
  const sandbox = {
    console,
    require,
    process,
    Buffer,
    setTimeout,
    setInterval,
    clearTimeout,
    clearInterval,
    __filename: sourcePath,
    __dirname: path.dirname(sourcePath),
  };
  const preExistingKeys = new Set(Object.keys(sandbox));
  const context = vm.createContext(sandbox);

  let exitCode = 0;
  try {
    vm.runInContext(source, context, { filename: sourcePath });
  } catch (err) {
    process.stderr.write((err && err.stack ? err.stack : String(err)) + '\n');
    exitCode = 1;
  } finally {
    const userKeys = Object.keys(sandbox).filter((k) => !preExistingKeys.has(k));
    try {
      fs.writeFileSync(VAR_DUMP_PATH, JSON.stringify(dumpContext(sandbox, userKeys)));
    } catch (dumpErr) {
      process.stderr.write(`kubesandbox_runner: failed to write variable dump: ${dumpErr}\n`);
    }
  }

  return exitCode;
}

process.exit(main());
